lldb/source/Target/SwiftLanguageRuntime.cpp

```diff
@@ -135,79 +135,98 @@ static bool HasReflectionInfo(ObjectFile *obj_file) {
-  reflection_ctx.reset(new NativeReflectionContext(this->GetMemoryReader()));
+  std::lock_guard<std::recursive_mutex> lock(m_add_module_mutex);
+
+  m_reflection_ctx.reset(new NativeReflectionContext(this->GetMemoryReader()));
+  m_initialized_reflection_ctx = true;
 
   auto &target = m_process->GetTarget();
   auto exe_module = target.GetExecutableModule();
-  if (!exe_module)
-    return;
-  auto *obj_file = exe_module->GetObjectFile();
-  if (!obj_file)
-      return;
-  if (obj_file->GetPluginName().GetStringRef().equals("elf"))
-    return;
-  Address start_address = obj_file->GetBaseAddress();
-  auto load_ptr = static_cast<uintptr_t>(start_address.GetLoadAddress(&target));
-
-  // Bail out if we can't read the executable instead of crashing.
-  if (load_ptr == 0 || load_ptr == LLDB_INVALID_ADDRESS)
+  if (!AddModuleToReflectionContext(exe_module)) {
+    m_reflection_ctx.reset();
     return;
+  }
 
-  reflection_ctx.reset(new NativeReflectionContext(this->GetMemoryReader()));
-  reflection_ctx->addImage(swift::remote::RemoteAddress(load_ptr));
+  m_modules_to_add.AppendIfNeeded(GetTargetRef().GetImages());
 
-  auto module_list = GetTargetRef().GetImages();
-  module_list.ForEach([&](const ModuleSP &module_sp) -> bool {
-    auto *obj_file = module_sp->GetObjectFile();
-    if (!obj_file)
-        return false;
-    if (obj_file->GetPluginName().GetStringRef().equals("elf"))
-      return true;
-    Address start_address = obj_file->GetBaseAddress();
-    auto load_ptr = static_cast<uintptr_t>(
-        start_address.GetLoadAddress(&(m_process->GetTarget())));
-    if (load_ptr == 0 || load_ptr == LLDB_INVALID_ADDRESS)
-      return false;
-    if (HasReflectionInfo(obj_file))
-      reflection_ctx->addImage(swift::remote::RemoteAddress(load_ptr));
+  // Add all defered modules to reflection context that were added to
+  // the target since this SwiftLanguageRuntime was created.
+  m_modules_to_add.ForEach([&](const ModuleSP &module_sp) -> bool {
+    AddModuleToReflectionContext(module_sp);
     return true;
   });
+  m_modules_to_add.Clear();
+
+  // The global ABI bit is read by the Swift runtime library.
+  SetupABIBit();
 }
 
 SwiftLanguageRuntime::SwiftLanguageRuntime(Process *process)
     : LanguageRuntime(process) {
-  SetupSwiftError();
-  SetupExclusivity();
-  SetupReflection();
-  SetupABIBit();
 }
 
 bool SwiftLanguageRuntime::IsABIStable() {
+  SetupABIBit();
   return _swift_classIsSwiftMask == 2;
 }
 
+enum class RuntimeKind { Swift, ObjC };
 static llvm::Optional<lldb::addr_t>
-FindSymbolForSwiftObject(Target &target, ConstString object,
-                         const SymbolType sym_type) {
-  llvm::Optional<lldb::addr_t> retval;
+FindSymbolForSwiftObject(Target &target, RuntimeKind runtime_kind,
+                         StringRef object, const SymbolType sym_type) {
+  StringRef filter((runtime_kind == RuntimeKind::Swift) ? "libswiftCore"
+                                                        : "libobjc");
+
```

Comment:
ObjCLanguageRuntime has a static method IsModuleObjCLibrary.  Probably better to use that - and the SwiftLanguageRuntime should have an equivalent if it doesn't already.